src/main/scala/bench/Benchmark.scala

```diff
@@ -0,0 +1,239 @@
+package bench
+
+import strawman.collection.{immutable, mutable}
+import strawman.collection.immutable.{LazyList, List}
+import java.lang.{Object, String}
+
+import org.scalameter.picklers.Implicits.stringPickler
+import org.scalameter._
+import strawman.collection.mutable.{ArrayBuffer, ListBuffer}
+
+import scala.{Any, Int, Unit}
+import scala.Predef.identity
+
+// You can run the benchmarks with `sbt run`
+// Comment the following `include(…)` lines to exclude a benchmark in particular
+// These benchmarks have been heavily inspired by those: https://github.com/lihaoyi/scala-bench
+object Benchmark extends Bench.Group {
+
+  include(new Cons {})
+  include(new Uncons {})
+  include(new Concat {})
+  include(new Foreach {})
+  include(new Lookup {})
+  include(new MemoryFootprint {})
+
+}
+
+trait MemoryFootprint extends Bench.ForkedTime with Generators {
+
+  override def measurer = new Executor.Measurer.MemoryFootprint
+
+  def benchmark[A](name: String, generator: Gen[A]): Unit =
+    performance of name in {
+      measure method "memory-footprint" in {
+        using(generator).curve("memory-footprint") in identity
+      }
+    }
+
+  benchmark("scala.List", scalaLists)
+  benchmark("List", lists)
+  benchmark("LazyList", lazyLists)
+  benchmark("ArrayBuffer", arrayBuffers)
+  benchmark("ListBuffer", listBuffers)
+
+}
+
+trait Cons extends Bench.ForkedTime with Generators {
+
+  def benchmark[A](name: String)(block: ((Int, Object)) => Any): Unit =
+    performance of name in {
+      measure method "cons" in {
+        using(sizeAndObjs).curve("cons") in block
+      }
+    }
+
+  benchmark("scala.List") { case (size, obj) =>
+    var xs = scala.List.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs = obj :: xs
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("List") { case (size, obj) =>
+    var xs = List.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs = obj :: xs
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("LazyList") { case (size, obj) =>
+    var xs = LazyList.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs = obj #:: xs
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("ArrayBuffer") { case (size, obj) =>
+    val xs = mutable.ArrayBuffer.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs += obj
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("ListBuffer") { case (size, obj) =>
+    val xs = mutable.ListBuffer.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs += obj
+      i += 1
+    }
+    xs
+  }
+
+}
+
+trait Uncons extends Bench.ForkedTime with Generators {
+
+  def benchmark[A](name: String, generator: Gen[A])(block: A => Any): Unit =
+    performance of name in {
+      measure method "tail" in {
+        using(generator).curve("uncons") in block
+      }
+    }
+
+  benchmark("scala.List", scalaLists)(_.tail)
+  benchmark("List", lists)(_.tail)
```

Comment:
Does a  single `tail` take enough time to be measurable?  I rarely get good results on single ns-length operations even with JMH which arguably tries the hardest to make this work.  You have to be careful with  `ArrayBuffer` which will be `O(n)`, but I think this may need a little more thought.